Don't use tokio::spawn

klochowicz opened the pull request:
see commit descriptions for more info

klochowicz committed:
Don't use tokio feature from xtra

Prefer using our FutureExt to enforce keeping RemoteHandles to prevent zombie
tasks.

klochowicz committed:
Tie lifetimes of spawned tasks to actors

klochowicz committed:
Disallow calls to tokio::spawn outside dedicated extension trait

Prevent accidentally spawning a task that could outlive the actor system
and become a zombie.

Any long-lived task should be tied to its owner by keeping RemoteHandle around.

thomaseizinger commented on daemon/src/connection.rs:
```diff
@@ -15,7 +15,7 @@ use xtra_productivity::xtra_productivity;
 struct ConnectedState {
     last_heartbeat: SystemTime,
-    _pulse_handle: RemoteHandle<()>,
+    tasks: Vec<RemoteHandle<()>>,
```
What do you think of using the `Tasks` struct in more places and extend it with an API like:

```rust
impl Tasks {
	pub fn add(&mut self, f: impl Future) {
		let handle = f.spawn_with_handle();
		self.0.push(handle);
	}
}
```

Then we could do things like:

```rust
tasks.add(self.send_to_maker_ctx.attach(send_to_socket));
```

Which would be a bit shorter.

klochowicz replied:
great idea!

klochowicz replied:
Added the API, guarded by a `debug_assert!` to prevent calling `spawn_with_handle()` twice (as it makes no sense). Guard is covered with a test, and it was helpful when refactoring the code (I missed 2!)

thomaseizinger commented on daemon/src/maker_cfd.rs:
```diff
@@ -765,18 +770,19 @@ where
+        let task = async move {
+            let dlc = contract_future.await;
+
+            log_error!(this.do_send_async(CfdRollOverCompleted { order_id, dlc }));
```
I don't think we should `log_error` here, this only ever fails with `Disconnected` which is unlikely (impossible?) because we are sending the message to ourselves.

klochowicz replied:
it was possible to get this one if the actor got shut down whilst doing contract setup (which takes ~1 min). With the guard that I've added here, it shouldn't ever happen, so I'm happy to do an `.expect("actor to be connected")` here. 

I'm not fussed, I just didn't want to swallow the error silently, and I also didn't want to open the doors for anything else than `RemoteHandle<()>`

klochowicz replied:
used `expect()` (as the master branch does now too)

thomaseizinger commented on daemon/src/maker_cfd.rs:
```diff
@@ -603,17 +607,18 @@ where
-        });
+            log_error!(this.do_send_async(CfdSetupCompleted { order_id, dlc }))
+        }
+        .spawn_with_handle();
```
I think this is unnecessary, this task will end by itself sooner or later. I think the idea of using `spawn_with_handle` should only be applied to tasks that contain endless loops.

klochowicz replied:
I agree that main use-case is dealing with tasks containing endless loops, however in this case cleanup during shutdown could deal with it in a nice way IMHO (especially that we already have a state for in-flight setup, we're making the code clearer this way for a reader)

thomaseizinger reviewed (commented):
Nice, thank you!

Some ideas to mull over.

klochowicz committed:
Expand Tasks API for ergonomics

The fact that we're storing a Vec<RemoteHandle<()>) internally is an
implementation detail.
Calling `spawn_with_handle` is done internally now, guarded by a `debug_assert!`
macro discouraging from doing it twice.

thomaseizinger commented on daemon/Cargo.toml:
```diff
@@ -45,7 +45,7 @@ tracing = { version = "0.1" }
 uuid = { version = "0.8", features = ["serde", "v4"] }
 x25519-dalek = { version = "1.1" }
-xtra = { version = "0.6", features = ["with-tokio-1"] }
+xtra = { version = "0.6" }
```
I believe this patch by itself will fail to compile yes? It should probably come after you've migrated all usages of `spawn_global`.

klochowicz replied:
It didn't fail to compile. We were not using `spawn_global` before this changeset. I moved it after as it didn't matter.

thomaseizinger reviewed (changes requested):
Commit message of 7723cb2174db1dd3fe87cf30df92b932cb0bc203 is a bit odd, is that on purpose?

One nit regard the order of the commit messages!

thomaseizinger commented on daemon/src/tokio_ext.rs:
```diff
@@ -39,12 +40,46 @@ where
+            result.is_err(),
+            "Spawning a remote handle into a separate task should panic"
+        );
+    }
```
This test will fail if we build the tests with `--release`. I think that is fine for now but perhaps we should just replace the `debug_assert` with an actual assert?

klochowicz replied:
thanks for pointing it out! I've changed the test to only panic in debug:

```rust
+        if cfg!(debug_assertions) {
+            assert!(
+                result.is_err(),
+                "Spawning a remote handle into a separate task should panic_in_debug_mode"
+            );
+        } else {
+            assert!(result.is_ok(), "Do not panic in release mode");
+        }
```
